### ISSUE TITLE
vmq_bridge plugin stops working after some failed tries because of invalid credentials

### ISSUE THREAD

**Issue by cdro-idirect** [vernemq/vernemq#1522]
Hi! I just ran into the follwing: I have 2 vernemq instances, one of them should open a ssl bridge to the other one. I accidently configured the wrong credentials on the remote broker, so I got some of the following messages:

```
2020-05-27 10:59:58.286 [error] <0.543.0>@vmq_bridge:on_connect_error:73 connection failed due to invalid_credentials
2020-05-27 11:00:08.363 [error] <0.559.0>@vmq_bridge:on_connect_error:73 connection failed due to invalid_credentials
2020-05-27 11:00:18.411 [error] <0.559.0>@vmq_bridge:on_connect_error:73 connection failed due to invalid_credentials
```

then I just get some error messages and the bridge shuts down.

```
2020-05-27 11:00:28.616 [error] <0.611.0>@gen_mqtt_client:active_once:745 gen_fsm <0.611.0> in state waiting_for_connack terminated with reason: no match of right hand value {error,closed} in gen_mqtt_client:active_once/2 line 745
2020-05-27 11:00:28.616 [error] <0.611.0>@gen_mqtt_client:active_once:745 CRASH REPORT Process <0.611.0> with 2 neighbours exited with reason: no match of right hand value {error,closed} in gen_mqtt_client:active_once/2 line 745 in gen_fsm:terminate/8 line 623
2020-05-27 11:00:28.617 [error] <0.468.0>@gen_mqtt_client:active_once:745 Supervisor vmq_bridge_sup had child {vmq_bridge,"10.68.10.201",8883} started with vmq_bridge:start_link(ssl, "10.68.10.201", 8883, {vmq_reg,direct_plugin_exports,[vmq_bridge]}, [{cleansession,false},{client_id,"verne38.sbr0"},{keepalive_interval,10},{restart_timeout,10},{username,...},...]) at <0.608.0> exit with reason no match of right hand value {error,closed} in gen_mqtt_client:active_once/2 line 745 in context child_terminated
2020-05-27 11:00:28.617 [error] <0.468.0> Supervisor vmq_bridge_sup had child {vmq_bridge,"10.68.10.201",8883} started with vmq_bridge:start_link(ssl, "10.68.10.201", 8883, {vmq_reg,direct_plugin_exports,[vmq_bridge]}, [{cleansession,false},{client_id,"verne38.sbr0"},{keepalive_interval,10},{restart_timeout,10},{username,...},...]) at <0.608.0> exit with reason reached_max_restart_intensity in context shutdown
2020-05-27 11:00:28.618 [info] <0.43.0> Application vmq_bridge exited with reason: shutdown
```
Even if I correct the credentials in the remote broker it does not start / try anymore.

### Environment

- VerneMQ Version: 1.10.2
- OS: CentOS 8

### Expected behavior

I don't know, maybe vernemq keeps trying to connect in a specific interval, indefinitely?

**Comment by ioolkos** [comment]
@cdro-idirect thanks, it must be an SSL error. The connecting process waits for a packet in active_once and then gets an {error, closed}, which terminates the process.

Is there anything notable in your SSL setup? what does the remote broker log?

**Comment by cdro-idirect** [comment]
Yes, I also assume that it has to do something with ssl, because if I don't use ssl that doesn't happen. 
There is nothing really special in my setup,
remote broker has this configured:
```
listener.ssl.default = 10.68.10.201:8883
listener.ssl.cafile = /etc/vernemq/cacert.crt
listener.ssl.keyfile = /etc/vernemq/mqtt-broker.key
listener.ssl.certfile = /etc/vernemq/mqtt-broker.crt
```

the broker that is opening the bridge has this configured:
```
vmq_bridge.ssl.sbr0 = 10.68.10.201:8883
vmq_bridge.ssl.sbr0.username = test
vmq_bridge.ssl.sbr0.password = test
vmq_bridge.ssl.sbr0.cafile = /etc/vernemq/cacert.crt
vmq_bridge.ssl.sbr0.certfile = /etc/vernemq/client.crt
vmq_bridge.ssl.sbr0.keyfile = /etc/vernemq/client.key
vmq_bridge.ssl.sbr0.topic.1 = test in 1
vmq_bridge.ssl.sbr0.restart_timeout = 10
vmq_bridge.ssl.sbr0.keepalive_interval = 10
```

The remote broker just logs
```
2020-05-27 10:59:58.284 [warning] <0.491.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:55765 due to invalid_credentials
2020-05-27 11:00:08.325 [warning] <0.499.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:57617 due to invalid_credentials
2020-05-27 11:00:08.361 [warning] <0.504.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:40549 due to invalid_credentials
2020-05-27 11:00:18.409 [warning] <0.512.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:32841 due to invalid_credentials
2020-05-27 11:00:28.452 [warning] <0.520.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:40639 due to invalid_credentials
2020-05-27 11:00:28.483 [warning] <0.525.0>@vmq_mqtt_fsm:check_user:567 can't authenticate client {[],<<"verne38.sbr0">>} from 10.68.10.203:48901 due to invalid_credentials
```

It is also not 100% reproducible, so it might indeed be something else. I would probably rule out a network issue, because the brokers were the above error happend, are running on my laptop in 2 vagrant boxes, connected via a network bridge.

**Comment by ioolkos** [comment]
can you set a username and password in the bridge config, and allow those at the remote broker?

**Comment by ioolkos** [comment]
Apologies, you seem to have set that, overlooked it. And those username/pw are actually really set in your auth plugin? (what do you use as the auth plugin btw).

**Comment by cdro-idirect** [comment]
I'm using vmq_passwd. I have configured "test:<hash>" in vmq.passwd file. The password is wrong on purpose, to recreate the issue. When I use the correct password the bridge connect without issues.

**Comment by ioolkos** [comment]
Ok, I think I understand what you mean. Your question is why the bridge crashes after 5 wrong connect attempts. 
And while I hope this is not a use case for you :) you might be right questioning it. Maybe the remote broker shouldn't just RST the client here, but do a slower disconnect handshake. (it's configurable)

**Comment by ioolkos** [comment]
If you do exactly as described in the linked comment (on the remote broker), you could verify the disconnect thingy:

https://github.com/vernemq/vernemq/issues/1409#issuecomment-574099245
The essential line is `{linger, {true, 5}}`.

**Comment by cdro-idirect** [comment]
Thanks. I tried that, but it didn't really help. I also made a capture with tcpdump. It seems, sometimes the remote brokers sends a RST,ACK to the client and the client sends an ACK for that, which doesn't make much sense guess. I will attach the capture here, you can find one tcp connection where it happens with this filter in wireshark: "tcp.stream eq 5", the connection starts at 15:49:38, remote broker is 10.68.10.201, client broker is 10.68.10.203, each one is running in a vagrant virtual machine on my laptop, connected by a network bridge.
[mqtt_bridge.zip](https://github.com/vernemq/vernemq/files/4694837/mqtt_bridge.zip)

**Comment by ioolkos** [comment]
Hm, okay. What happens if you set `exit_on_close` to false, in the advanced.config? Don't forget to restart the node.

**Comment by cdro-idirect** [comment]
This doesn't change anything. I still see these "CRASH REPORT" error messages and sooner or later the bridge plugin will shutdown.

**Comment by ioolkos** [comment]
Ok. What makes this interesting is that there's TLS, TCP and MQTT involved: the server finds an invalid password in the MQTT CONNECT packet and has to terminate the connection. 
My idea here was to try to make the server not RST the connection to see whether this would change anything or not. (only Wireshark would show if there's still an RST though)

**Comment by cdro-idirect** [comment]
Created a new capture file, this time with the advanced.config and exit_on_close set to false and left it running until the bridge plugin stopped working (took around 19 minutes). 
When looking at the capture, I noticed that especially for the last connection attempt on 15:19:42 there were actually 4 connection attempts within a couple milliseconds. Immediately after getting the RST, the clients tries to reconnect.
[mqtt_bridge2.zip](https://github.com/vernemq/vernemq/files/4696615/mqtt_bridge2.zip)

**Comment by ioolkos** [comment]
Thanks, I do not understand your remark about the bridge working for 19 minutes.It should not connect at all if the TLS and/or authentication doesn't work.

**Comment by cdro-idirect** [comment]
Yes, what I meant was "stopped trying to connect". So for 19 minutes it tries to connect, and then it stopped doing that.

**Comment by ioolkos** [comment]
Okay, that's funny. This probably just flies under the radar for the restart_exhaustion. 
And then once it's triggered the bridge app supervisor obviously terminates the bridge.

2 action points:
 - is there no runtime error, when target sends `FIN` instead of `RST` (above tests tried and failed to generate `FIN`, it seems)
 - secure the local socket option setting (setting to `active once`) against the remote TLS socket massacre
